homeassistant/components/broadlink/device.py

```diff
@@ -29,18 +38,144 @@ def get_domains(device_type):
     return {d for d, t in DOMAINS_AND_TYPES.items() if device_type in t}
 
 
+class BroadlinkStores:
+    """Manages a storage setup for a device."""
+
+    def __init__(self, codes_store: Store, flags_store: Store) -> None:
+        """Initialize stores."""
+        self._code_storage = codes_store
+        self._flag_storage = flags_store
+        self._codes: dict[str, dict[str, str | list[str]]] = {}
+        self._flags: dict[str, int] = defaultdict(int)
+        self._storage_loaded = False
+
+    async def async_setup(self) -> None:
+        """Load all stores and data."""
+        self._codes.update(await self._code_storage.async_load() or {})
+        self._flags.update(await self._flag_storage.async_load() or {})
+
+    def extract_devices_and_commands(self) -> dict[str, list[str]]:
+        """Return the set of devices and commands in storage."""
+        return {device: list(subdevices) for device, subdevices in self._codes.items()}
+
+    def extract_codes(
+        self, commands: list[str], device: str | None = None
+    ) -> list[list[bytes]]:
+        """Extract a list of codes.
+
+        If the command starts with `b64:`, extract the code from it.
+        Otherwise, extract the code from storage, using the command and
+        device as keys.
+
+        The codes are returned in sublists. For toggle commands, the
+        sublist contains two codes that must be sent alternately with
+        each call.
+        """
+
+        def _data_packet_annotated(code):
+            try:
+                return data_packet(code)
+            except ValueError as err:
+                raise ValueError(f"Invalid code: {repr(code)}") from err
+
+        code_list = []
+        for cmd in commands:
+            if cmd.startswith("b64:"):
+                codes = [cmd[4:]]
+
+            else:
+                if device is None:
+                    raise ValueError("You need to specify a device")
+
+                try:
+                    codes_raw = self._codes[device][cmd]
+                except KeyError as err:
+                    raise ValueError(f"Command not found: {repr(cmd)}") from err
+
+                if isinstance(codes_raw, list):
+                    codes = codes_raw[:]
+                else:
+                    codes = [codes_raw]
+
+            data = [_data_packet_annotated(code) for code in codes]
+
+            code_list.append(data)
+        return code_list
+
+    def toggled_codes(
+        self, code_list: list[list[bytes]], subdevice: str | None = None
+    ) -> Generator[bytes, None, None]:
+        """Generate the list of codes we want and toggle as we go along."""
+        try:
+            for codes in code_list:
+                if len(codes) > 1 and subdevice:
+                    yield codes[self._flags[subdevice]]
+                    self._flags[subdevice] ^= 1
```

Comment:
We can only change the flag when the request is successful.

Reply:
The yield will stop execution untill the next item is requested, at which point the code has been transmitted.